### PR TITLE
Support binary path in Protocol.extract_impls/2

### PR DESCRIPTION
**Pull request by kpanic**
How to reproduce:

This works:

```elixir
    iex> path = :code.lib_dir(:elixir, :ebin)
    iex> mods = Protocol.extract_impls(Enumerable, [path])
    iex> List in mods
    true
```

This does not work

```elixir
    iex> path = Application.app_dir(:elixir, "ebin")
    iex> mods = Protocol.extract_impls(Enumerable, [path])
    iex> List in mods
    false
```

However the spec says:

```elixir
iex(1)> h Protocol.extract_impls

                       def extract_impls(protocol, paths)

  @spec extract_impls(module(), [charlist() | String.t()]) :: [atom()]
```

And it accepts also a `String.t()`

### CODE REVIEW

File: lib/elixir/lib/protocol.ex [hunk 1]
```diff
@@ -474,6 +474,8 @@ defmodule Protocol do
   defp list_dir(path), do: list_dir(to_charlist(path))
 
   defp extract_from_file(path, file, prefix, callback) do
+    path = to_charlist(path)
```
Comment (josevalim) [inline lib/elixir/lib/protocol.ex]: Given that we also convert `to_charlist` inside `list_dir`, I would rather change the comprehension a couple lines above to this:

```
for path <- paths,
    path = to_charlist(path),
    file <- ...
```

Comment (josevalim) [reply lib/elixir/lib/protocol.ex]: And then remove the `to_charlist` call on both `list_dir` and `extract_from_file`.

Comment (kpanic) [reply lib/elixir/lib/protocol.ex]: Done

**Comment by ericmj** [comment]
Thank you! 💜

File: lib/elixir/test/elixir/protocol/consolidation_test.exs [hunk 2]
```diff
@@ -149,12 +149,18 @@ defmodule Protocol.ConsolidationTest do
...
   end
 
-  test "consolidation extracts implementations" do
+  test "consolidation extracts implementations with charlist path" do
     protos = Protocol.extract_impls(Enumerable, [:code.lib_dir(:elixir, :ebin)])
     assert List in protos
     assert Function in protos
   end
 
+  test "consolidation extracts implementations with binary path" do
```
Comment (lorenzosinisi) [inline lib/elixir/test/elixir/protocol/consolidation_test.exs]: 💯

**Pull request merged**